### PR TITLE
[FLINK-6725][table] make requiresOver as a contracted method in udagg

### PR DESCRIPTION
**Pull request by shaoxuan-wang**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by fhueske** [comment]
The change itself is OK. However, I added a comment about breaking the API and the change itself to the corresponding JIRA issue.

**Comment by shaoxuan-wang** [comment]
Thanks for the review @fhueske @sunjincheng121 . 
Fabian, yes, I do not think this will be a blocker for release. In fact, this is a similar interface clean-up issue as FLINK-6457 that we want to solve. But certainly, we can just merge this to master and include this in the next release.

Jincheng, yes, it is one's flavor about how to implement an interface with contracted method. From my point of view:
1. A contract method is usually to be very useful as a flag to indicate the characteristic of an aggregator, and this information will help the optimizer to decide the plan. Such methods are retract/merge/resetAccumulator.
2. I prefer we should just expose the user interface without any default implementation (an interface method with default implementation will anyway not control the user behavior). The user defined interface IMO should purely ask users to provide minimum required methods that can let the UDX work. 
3. We will call getAccumulatorType and getResultType also during the compilation to generate the plan I think. If we put requiresOver into the interface, we should also consider to put getAccumulatorType and getResultType into interface as well.

Also, I think we may want to provide all the agg that requiresOver as the built-in agg in the future. Thereby, we may not need to expose this interface in the UDAGG at that point.

**Comment by fhueske** [comment]
Hi @shaoxuan-wang and @sunjincheng121,

Ad 2) I think it is OK to have this as a method with a default implementation. If we choose the contract function design, we have the same default behavior as the default implementation if the method is not implemented, so IMO there is not really a difference. Due to the default implementation, a user does not need to implement the method unless it is necessary. However in that case, it can be overridden in a safe way with IDE support. A contract function, does not offer this safety because a user needs to lookup the exact signature in the documentation.
Ad 3) I think it would indeed make sense to add `getAccumulatorType()` and `getResultType()` to `AggregateFunction` and return `null` by default. This would also be more consistent with `ScalarFunction` and `TableFunction`. This would also help with type and compile safety, because we can enforce the type of the returned `TypeInformation` as `T` and `ACC` without going through reflection magic.

**Comment by sunjincheng121** [comment]
Hi  @shaoxuan-wang, the suggestions of @fhueske  make sense to me.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 1]
```diff
@@ -329,6 +330,21 @@ object UserDefinedFunctionUtils {
   }
 
   /**
+    * Returns the value (boolean) of AggregateFunction#requiresOver() that indicates if this
+    * AggregateFunction can only be used for over window aggregate. If method requiresOver is not
+    * provided, return false as default value.
+    */
+  def getRequiresOverConfig(aggregateFunction: AggregateFunction[_, _]): Boolean = {
+    try {
+      val method: Method = aggregateFunction.getClass.getMethod("requiresOver")
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: Yes, I think we should check the return type of `requiresOver` as well , in order to give an explicit exception.

**Comment by wuchong** [comment]
Regarding to (3), I agree with Fabian. From a user's perspective, it is not easy to implement a `getAccumulatorType()` or `getResultType()` contract function, as the signature and return type is complicated.

**Comment by fhueske** [comment]
How do we continue with this PR @shaoxuan-wang?

**Comment by shaoxuan-wang** [comment]
@fhueske @sunjincheng121 @wuchong, thanks for the valuable inputs. We talked offline and get an agreement that we'd better to put `getAccumulatorType()`, `getResultType()` etc. in `AggregateFunction`. Sorry that I forgot to update this Jira. Let us move this change to [FLINK-7194](https://issues.apache.org/jira/browse/FLINK-7194). I will close this PR and [FLINK-6725](https://issues.apache.org/jira/browse/FLINK-6725).

**Comment by fhueske** [comment]
No worries. Thanks for the update!

**Pull request closed without merging**